horovod/run/elastic/driver.py

```diff
@@ -0,0 +1,283 @@
+                    .format(min_np=min_np, extra_message=extra_message))
+
+        self._wait_hosts_cond.acquire()
+        try:
+            while True:
+                current_hosts = self._host_manager.current_hosts
+                if current_hosts.count_available_slots() >= min_np and len(current_hosts.available_hosts) >= min_hosts:
+                    return current_hosts
+                self._wait_hosts_cond.wait(tmout.remaining())
+                tmout.check_time_out_for('minimum number of hosts to become available')
+        finally:
+            self._wait_hosts_cond.release()
+
+    def _activate_hosts(self, min_np):
+        logging.info('wait for available hosts: {}'.format(min_np))
+        current_hosts = self.wait_for_available_hosts(min_np)
+        pending_slots = self._update_host_assignments(current_hosts)
+        self._worker_registry.reset(self.world_size())
+        self._start_worker_processes(pending_slots)
+
+    def _discover_hosts(self):
+        first_update = True
+        while not self._shutdown.is_set():
+            self._wait_hosts_cond.acquire()
+            try:
+                if self._host_manager.update_available_hosts():
+                    self._notify_workers_host_changes(self._host_manager.current_hosts)
+                    self._wait_hosts_cond.notify_all()
+            except RuntimeError as e:
+                if first_update:
+                    # Misconfiguration, fail the job immediately
+                    raise
+                # Transient error, retry until timeout
+                logging.warning(str(e))
+            finally:
+                self._wait_hosts_cond.release()
+            first_update = False
+            self._shutdown.wait(DISCOVER_HOSTS_FREQUENCY_SECS)
+
+    def _notify_workers_host_changes(self, current_hosts):
+        # Assignments are required to be stable via contract
+        next_host_assignments, _ = self._get_host_assignments(current_hosts)
+        if next_host_assignments == self._host_assignments:
+            # Skip notifying workers when host changes would not result in changes of host assignments
+            logging.debug('no host changes, skipping notifications')
+            return
+
+        coordinator_slot_info = self._rank_assignments.get(0)
+        if not coordinator_slot_info:
+            logging.debug('no coordinator info, skipping notifications')
+            return
+
+        coordinator_client = self._worker_clients.get((coordinator_slot_info.hostname,
+                                                       coordinator_slot_info.local_rank))
+        if not coordinator_client:
+            logging.debug('no coordinator client, skipping notifications')
+            return
+
+        timestamp = _epoch_time_s()
+        try:
+            coordinator_client.notify_hosts_updated(timestamp)
+        except:
+            if self._verbose >= 2:
+                logging.exception('failed to notify {}[{}] of host updates'
```

Comment:
the exception message might be useful here, given we bothered setting `verbose=2`

Reply:
`logging.exception` will also print out the exception that was caught along with whatever else you provide.